File: modules/tools/exercises/06_multi_tool/solution/python/main.py
import json, os, urllib.request
def main():
    base=os.environ.get("MISTRAL_API_BASE","https://api.mistral.ai/v1").rstrip("/")
    payload={"model":"mistral-small-latest","messages":[{"role":"user","content":"search docs for RAG"}],
             "tools":[{"type":"function","function":{"name":"search"}},{"type":"function","function":{"name":"calculator"}}]}
    req=urllib.request.Request(f"{base}/chat/completions",data=json.dumps(payload).encode(),method="POST")
    req.add_header("Content-Type","application/json")
    with urllib.request.urlopen(req,timeout=30) as r: print(json.loads(r.read())["choices"][0]["message"]["content"])
if __name__=="__main__": main()

File: modules/tools/exercises/06_multi_tool/solution/cpp/main.cpp
#include "codam_llm.hpp"
#include <iostream>

int main() {
    nlohmann::json messages = nlohmann::json::array({
        {{"role", "user"}, {"content", "search docs for RAG"}},
    });
    nlohmann::json tools = nlohmann::json::array({
        {{"type", "function"}, {"function", {{"name", "search"}}}},
        {{"type", "function"}, {"function", {{"name", "calculator"}}}},
    });
    std::cout << codam::chat_completion(messages, -1, {{"tools", tools}}) << "\n";
    return 0;
}
